src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,453 @@
+
+    spdlog::trace("considered config domain={} function={} injectionType={} probability={}",
+        domain, cbInfo->functionName, injectionType, injectionProbability);
+    if (injectionProbability < 100) {
+        if (injectionProbability <= 0) {
+            return;
+        }
+        const int rand10000 = rand() % 10000;
+        const int skipThreshold = injectionProbability * 10000 / 100;
+        spdlog::trace("rand1000={} skipThreshold={}", rand10000, skipThreshold);
+        if (rand10000 >= skipThreshold) {
+            return;
+        }
+        spdlog::debug(
+            "matched config based on rand10000={} skipThreshold={} "
+            "domain={} function={} injectionType={} probability={}",
+            rand10000, skipThreshold, domain,
+            cbInfo->functionName, injectionType, injectionProbability);
+    } else {
+        spdlog::debug("matched 100% config domain={} function={} injectionType={} probability={}",
+            domain, cbInfo->functionName, injectionType, injectionProbability);
+    }
+
+    switch (injectionType)
+    {
+    case FI_TRAP:
+        faultInjectorKernelTrap<<<1,1>>>();
+        break;
+
+    case FI_ASSERT:
+        faultInjectorKernelAssert<<<1,1>>>();
+        break;
+
+    case FI_RETURN_VALUE:
+        if (domain == CUPTI_CB_DOMAIN_DRIVER_API) {
+            CUresult *cuResPtr = static_cast<CUresult*>(cbInfo->functionReturnValue);
+            *cuResPtr = static_cast<CUresult>(substituteReturnCode);
+        } else if (domain == CUPTI_CB_DOMAIN_RUNTIME_API) {
+            cudaError_t *cudaErrPtr = static_cast<cudaError_t*>(cbInfo->functionReturnValue);
+            spdlog::error("updating runtime return value DOES NOT WORK, use trap or assert");
+            *cudaErrPtr = static_cast<cudaError_t>(substituteReturnCode);
+            break;
+        }
+
+    default:
+        break;
+    }
+}
+
+/**
+ * cuInit hook entry point
+ */
+int STDCALL
+InitializeInjection(void) {
+    spdlog::info("InitializeInjection");
+    if (globalControl.initialized) {
+        return 1;
+    }
+    // Init globalControl
+    globalControlInit();
+    globalControl.initialized = 1;
+
+    registerAtExitHandler();
+
+    if (globalControl.dynamic) {
+        spdlog::debug("creating a thread to watch the fault injector config interactively");
+        PTHREAD_CALL(pthread_create(&globalControl.dynamicThread, NULL, dynamicReconfig, NULL));
+    }
+
+    // Initialize CUPTI
+    CUPTI_CALL(cuptiInitialize());
+
+    return 1;
+}
+
+
+/**
+ * Parse and apply new config
+ */
+static void
+readFaultInjectorConfig(void) {
+    if (!configFilePath) {
+        spdlog::error("specify convig via environment {}", FAULT_INJECTOR_CONFIG_PATH);
+        return;
+    }
+    std::ifstream jsonStream(configFilePath);
+    if (!jsonStream.good()){
+        spdlog::error("check file exists {}", configFilePath);
+        return;
+    }
+
+    PTHREAD_CALL(pthread_rwlock_wrlock(&globalControl.configLock));
```

Comment:
Thank you for this bit: I had to go look up how we'd do this in C++17: [`std::shared_mutex`](https://en.cppreference.com/w/cpp/thread/shared_mutex), coupled with [`std::unique_lock`](https://en.cppreference.com/w/cpp/thread/unique_lock) for RAII.

Looks like this is doable with the standard library. We can address this later (should we choose to address it at all).